### ISSUE TITLE
Can't use separate cache configurations in an application

### ISSUE THREAD

**Issue by mjpieters** [long2ice/fastapi-cache#142]
While the `FastAPICache` object is implemented as a class, it is really just a singleton namespace; all attributes are class variables and all methods are class methods. All state it holds is global.

This means you can't define multiple cache configurations, hindering component-wise application construction.

E.g. if you wanted to store your RESTful API route cache in Redis, but use an in-memory cache for the HTML-heavy frontend portion of your application, you can't right now because there is just one global backend option. Nor can you set separate prefixes, unless you set a custom namespace for each route in a group of routes, something that's easy to get wrong.

The `FastAPICache` object should really be a real class. Instances could be used as the `cache` decorator, as well as let you adjust configuration with lifespan events:

```python
cache = FastAPICache(prefix="fastapi-cache")


@asynccontextmanager
async def lifespan(app: FastAPI):
    redis = aioredis.from_url("redis://localhost")
    cache.backend = RedisBackend(redis)
    yield


app = FastAPI(lifespan=lifespan)

@app.get("/")
@cache(expire=60)
async def index():
    return dict(hello="world")
```

The `cache` object then has all the context it needs to access the backend and shared configuration but additional `cache` `FastAPICache` instances can be created with different settings.

For backwards compatibility the current `cache` decorator could be using a global singleton instance of `FastAPICache()` but issue a deprecation warning.

**Comment by mkdir700** [comment]
Regarding compatibility, I have a question about whether some class methods in `FastAPICache`, such as `init` and `get_backend`, need to be backward compatible.

**Comment by mjpieters** [comment]
> Regarding compatibility, I have a question about whether some class methods in `FastAPICache`, such as `init` and `get_backend`, need to be backward compatible.

Yes, if we are making an attempt at backwards compatibility here then the classmethods need to be retained for now.

Note that some of my recent PRs have already broken BC in places; switching the backends and coder to using bytes will break any custom coders. There will have to be a 'porting to #nextversion' section in the documentation, I think.

**Comment by mkdir700** [comment]
The `@cache` decorator will use a default global instance of `FastAPICache`, which is temporarily named as `default_cacher`. When calling class methods such as `FastAPICache.get_backend`, `FastAPICache.get_coder`, and `FastAPICache.get_key_builder`, it actually accesses the properties related to the `default_cacher`.

Is my understanding above correct?

**Comment by mjpieters** [comment]
> Is my understanding above correct?

Exactly. And these would all emit a deprecation warning.

**Comment by mkdir700** [comment]
> 

I got it. Are you currently implementing this feature? I can try to submit a PR for it.

**Comment by elad-aharon** [comment]
Hi folks, I'll add my 2 cents to the discussion:

I'm using your cache lib in my application and I wanted to test it. I love the idea that cache is globally by default for the app, but not letting me using cache object it's annoying for this reasons:

- (main reason) tests reusing same cache class, if they are `async` it's a fertile ground for bugs. i.e: testA might cache response testB might use although didn't expect that.  
- application cannot separate caches i.e: some endpoints will use redis and some dynamodb.

any thoughts for workarounds / ETA for `default_cacher`?
I could also try PR it if needed with your guidance.
